### PR TITLE
Remove dependency on Boost.MPI and Boost.Serialization

### PR DESCRIPTION
**Pull request by krivenko**
Okay, this one is bigger and more involved.

* Boost types such as `boost::mpi::communicator` have been replaced with their plain MPI counterparts.
* New utility header `mpi_dispatcher/misc.hpp`.
* GF term structures got their own [MPI structure datatypes](https://pages.tacc.utexas.edu/~eijkhout/pcse/html/mpi-data.html#Structtype).
* Interface of `TermList` has been slightly extended.
* All unit tests and programs have been updated + one test added (`broadcast`).
* `prog/anderson.cpp` and `prog/hubbard2d.cpp` are gone as they were unused and broken.

An important question here is whether it is reasonable to ask for MPI-3.0 (introduced in September 2012).
I would like to have it because of the predefined MPI datatypes `MPI_CXX_BOOL` and `MPI_CXX_DOUBLE_COMPLEX`.

OpenMPI supports MPI-3.0 since version 1.8 (Mar 2014).
MPICH and derivatives support it since version 3.0.0 (Dec 2012).

If we decide that MPI-3.0 is too much to ask, I will write and push a workaround for those datatypes. Otherwise `.travis.yml` must be upgraded to use Ubuntu Xenial as Trusty ships only `libopenmpi1.6`.

### CODE REVIEW

**Comment by aeantipov** [comment]
Igor, I will review this, but before that - MPI-3.0 is fine imho.

**Comment by krivenko** [comment]
@aeantipov Understood. I'll fix the Travis config later this week.

File: src/pomerol/TwoParticleGFPart.cpp [hunk 1]
```diff
@@ -40,18 +38,37 @@ TwoParticleGFPart::NonResonantTerm& TwoParticleGFPart::NonResonantTerm::operator
     return *this;
 }
 
+MPI_Datatype TwoParticleGFPart::NonResonantTerm::mpi_datatype() {
```
Comment (aeantipov) [inline src/pomerol/TwoParticleGFPart.cpp:41]: Igor, mind writing a docstring on what this is? I realize it's a replacement to serialization.

**Review by aeantipov — APPROVED** [review]
This looks very good.

**Comment by krivenko** [comment]
I have updated Travis config, but the build fails due to an Eigen bug. Ubuntu maintainers apparently think that it is a good idea to package a beta version of Eigen as part of an LTS release.

Will need a bit of time to resolve that one.
https://gitlab.com/libeigen/eigen/-/issues/1359

**Comment by krivenko** [comment]
@aeantipov Please, check [the few changes](https://github.com/aeantipov/pomerol/pull/32/files/26d7d04766ca537e6035fa3980516f8e0a29f46b..8af8b4398305a37abcb129d307569485815c292e) I made since your approval.

I have updated the Travis CI configuration so that is uses the Ubuntu 16.04 image. I left only 4 compilers in the jobs matrix,

- `g++-5.0` (system default)
- `g++-4.9`
- `clang++-3.6`
- `clang++-3.7`

I suggest we merge this PR despite the fact that the last 3 compilers fail to link executables in the `prog` directory.

The source of the failures is described here: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=797917
In a nutshell, packages in Ubuntu Xenial -- including Boost -- are built using GCC 5, which generates shared libraries with a new version of C++ ABI (C++11). The older versions of GCC and Clang <4.0 still use the old ABI, which causes problems when one tries to link excutables to the new `libboost_program_options.so` library.

If we still want to support the older compilers (which, I believe, makes sense), we should consider removing those program_options bits and replacing them with a bundled tclap or [args.hxx](https://github.com/Taywee/args) in a later PR.

**Comment by aeantipov** [comment]
Sounds reasonable, let's add the issues we stumbled upon to keep track. Feel free to merge.

**Comment by aeantipov** [comment]
Reference #33
